apps/chat/tests/test_scheduled_messages.py

```diff
@@ -0,0 +1,320 @@
+        scheduled_message.refresh_from_db()
+        if expect_next_trigger_date_changed:
+            # Assert that the date actually moved on
+            assert scheduled_message.next_trigger_date > prev_next_trigger_date
+
+        assert scheduled_message.next_trigger_date == expected_next_trigger_date
+        assert scheduled_message.last_triggered_at == expected_last_triggered_at
+        assert scheduled_message.total_triggers == expected_total_triggers
+        assert scheduled_message.resolved == expected_resolved
+
+    session = ExperimentSessionFactory()
+    experiment = session.experiment
+    team = experiment.team
+    schedule_conf = ScheduledMessageConfigFactory(
+        experiment=experiment,
+        team=team,
+        time_period=TimePeriod(period),
+        frequency=1,
+        repetitions=2,
+    )
+    delta = relativedelta(**{schedule_conf.time_period: schedule_conf.frequency})
+    seconds_offset = 1
+    step_delta = delta + relativedelta(seconds=seconds_offset)
+
+    with freeze_time("2024-04-01") as frozen_time, patch("apps.chat.tasks.functions.Now") as db_time:
+        current_time = db_time.return_value = datetime.now().astimezone(pytz.timezone("UTC"))
+        scheduled_message = ScheduledMessageFactory(participant=session.participant, schedule=schedule_conf)
+        # Set the DB time to now
+
+        # DB is now behind the next trigger date
+        poll_scheduled_messages()
+        test_scheduled_message_attrs(
+            expected_next_trigger_date=current_time + delta,
+            expected_last_triggered_at=None,
+            expected_total_triggers=0,
+            expected_resolved=False,
+            expect_next_trigger_date_changed=False,
+        )
+
+        current_time = step_time(frozen_time, db_time, step_delta)
+        poll_scheduled_messages()
+        test_scheduled_message_attrs(
+            expected_next_trigger_date=current_time + delta,
+            expected_last_triggered_at=current_time,
+            expected_total_triggers=1,
+            expected_resolved=False,
+        )
+
+        current_time = step_time(frozen_time, db_time, step_delta)
+        poll_scheduled_messages()
+        # Since the scheduled message is be resolved, the expected_next_trigger_date will not be updated
+        test_scheduled_message_attrs(
+            # We subtract the offset here to account for it not being in the original `next_trigger_date`
+            expected_next_trigger_date=current_time - relativedelta(seconds=seconds_offset),
+            expected_last_triggered_at=current_time,
+            expected_total_triggers=2,
+            expected_resolved=True,
+            expect_next_trigger_date_changed=False,
+        )
+
+        # We are done now, but let's make 110% sure that another message will not be triggered
+        current_time = step_time(frozen_time, db_time, step_delta)
+        scheduled_message._trigger = Mock()
```

Comment:
I don't think this will work since the mock is on a different instance of the object than the one which will be returned from the DB in a query.

Maybe check that `_get_messages_to_fire` doesn't return any?

Reply:
Yeah you're right, not sure what I was thinking here tbh. Checking `_get_messages_to_fire` sounds better

Reply:
Changes are made [here](https://github.com/dimagi/open-chat-studio/pull/373/commits/6996e196dc707b432776eb8dc8b6312948cd14a4)